### PR TITLE
Set etag properly to a calculated value

### LINKED ISSUE #5100: server generates emtpy etag when client does not set content-md5
## Expected Behavior
Currently the server should compute the MD5 sum of the object and use this value as etag if the client does not set the content-md5. (Actually we can set a random value as etag - see #4385 #4641)  

fs.json
```

{
   "version":"1.0.1",
   "format":"fs","
    minio":{ 
             "release":"DEVELOPMENT.GOGET"
    },
   "meta":{
            "content-type":"application/octet-stream",
            "etag":"5eb63bbbe01eeed093cb22bb8f5acdc3"
   }
}
```

## Current Behavior
The server saves an empty value as etag. 

fs.json
```

{
   "version":"1.0.1",
   "format":"fs","
    minio":{ 
             "release":"DEVELOPMENT.GOGET"
    },
   "meta":{
            "content-type":"application/octet-stream",
            "etag":""
   }
}
```

## Possible Solution
The problem is within the changes of #5071. 
1. The `HashReader` should return the computed MD5 not the MD5 sum provided by the client (it may be emtpy) - See [MD5()](https://github.com/minio/minio/blob/master/pkg/hash/reader.go#L104)
2. The etag (metadata field) should be set after the client has uploaded the whole object (see Expected Behavior) - See  FS [Put](https://github.com/minio/minio/blob/master/cmd/fs-v1.go#L596)

## Context
Encrypted objects "don't have" an etag value any more. 

## Your Environment
current master

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Set etag properly to a calculated value

## Motivation and Context
Fixes https://github.com/minio/minio/issues/5100

## How Has This Been Tested?
Manually without setting `content-md5` 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/fs-v1.go [hunk 1]
```diff
@@ -593,7 +594,7 @@ func (fs fsObjects) PutObject(bucket string, object string, data *hash.Reader, m
 		return ObjectInfo{}, traceError(errInvalidArgument)
 	}
 
-	metadata["etag"] = data.MD5HexString()
```
Comment (aead) [inline cmd/fs-v1.go:596]: IMO we still have to move this line below the the `fsCreate` Otherwise the etag is always the same for PUTs without content-md5

Comment (harshavardhana) [reply cmd/fs-v1.go:596]: Done @aead

File: pkg/hash/reader.go [hunk 2]
```diff
@@ -52,13 +52,12 @@ func NewReader(src io.Reader, size int64, md5Hex, sha256Hex string) (*Reader, er
 	if err != nil {
 		return nil, err
 	}
+
 	var (
 		md5Hash    hash.Hash
```
Comment (aead) [inline pkg/hash/reader.go]: md5Hash is not required - we can directly use `md5Hash: md5.New()` during Reader init.

**Review by aead — CHANGES_REQUESTED** [review]
Everything else looks good to me 👍 
We have to revisit this again when we change etag with `-1` or `-part-number`.

File: pkg/hash/reader.go (hunk 2 above)
Comment (harshavardhana) [reply pkg/hash/reader.go]: Sure let me do that.

**Review by vadmeste — APPROVED** [review]
LGTM & Tested

**Review by aead — APPROVED** [review]
LGTM

**Pull request merged**